src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -126,14 +139,34 @@ internal partial class Binder
 
             return new BoundConversion(
                 syntax,
-                source,
+                BindToNaturalType(source, diagnostics),
                 conversion,
                 @checked: CheckOverflowAtRuntime,
                 explicitCastInCode: isCast && !wasCompilerGenerated,
                 conversionGroupOpt,
                 constantValueOpt: constantValue,
-                type: destination)
-            { WasCompilerGenerated = wasCompilerGenerated };
+                type: destination) { WasCompilerGenerated = wasCompilerGenerated };
+        }
+
+        /// <summary>
+        /// Rewrite the expressions in the switch expression arms to add a conversion to the destination type.
+        /// </summary>
+        private BoundExpression ConvertSwitchExpression(BoundUnconvertedSwitchExpression source, TypeSymbol destination, DiagnosticBag diagnostics, bool hasErrors = false)
+        {
+            var builder = ArrayBuilder<BoundSwitchExpressionArm>.GetInstance();
+            foreach (var oldCase in source.SwitchArms)
+            {
+                var oldValue = oldCase.Value;
+                var newValue = GenerateConversionForAssignment(destination, oldValue, diagnostics);
+                var newCase = (oldValue == newValue) ? oldCase :
+                    new BoundSwitchExpressionArm(oldCase.Syntax, oldCase.Locals, oldCase.Pattern, oldCase.WhenClause, newValue, oldCase.Label, oldCase.HasErrors);
+                builder.Add(newCase);
+            }
+
+            var newSwitchArms = builder.ToImmutableAndFree();
```

Comment:
Could consider an optimization where if all cases are the same, we don't need to create a new array.

Reply:
Someone once said "Aside: Man is this so much easier to understand" when I reverted such a micro-optimization.  I agree with that person.

---
In reply to: [301335232](https://github.com/dotnet/roslyn/pull/37052#discussion_r301335232) [](ancestors = 301335232)

Reply:
I don't think a bool flag for whether the builder should be used is quite so egregious

---
In reply to: [301743214](https://github.com/dotnet/roslyn/pull/37052#discussion_r301743214) [](ancestors = 301743214,301335232)